# Workspace Setup

In [53]:
import pandas as pd

df = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/lego-analysis/master/datasets/lego_sets.csv')
parent_theme = pd.read_csv('https://raw.githubusercontent.com/KeithGalli/lego-analysis/master/datasets/parent_themes.csv')
    


# Some Data Cleaning (not all)

In [259]:
df.head(3)

set_num                        name  year  num_parts    theme_name  \
0    00-1             Weetabix Castle  1970      471.0        Castle   
1  0011-2           Town Mini-Figures  1978        NaN  Supplemental   
2  0011-3  Castle 2 for 1 Bonus Offer  1987        NaN  Lion Knights   

  parent_theme  
0     Legoland  
1         Town  
2       Castle

In [258]:
parent_theme.head(3)

id     name  is_licensed
0   1  Technic        False
1  22  Creator        False
2  50     Town        False

In [255]:
merged.head(3)

set_num                        name_x  year  num_parts theme_name  \
0    00-1               Weetabix Castle  1970      471.0     Castle   
1    00-2  Weetabix Promotional House 1  1976        NaN   Building   
2    00-3  Weetabix Promotional House 2  1976        NaN   Building   

  parent_theme   id    name_y  is_licensed  
0     Legoland  411  Legoland        False  
1     Legoland  411  Legoland        False  
2     Legoland  411  Legoland        False

In [256]:
licensed.head(3)

set_num                       name_x  year  num_parts  \
3493  10018-1                   Darth Maul  2001     1868.0   
3494  10019-1  Rebel Blockade Runner - UCS  2001        NaN   
3495  10026-1      Naboo Starfighter - UCS  2002        NaN   

                   theme_name parent_theme   id  is_licensed  count  
3493                Star Wars    Star Wars  158         True      1  
3494  Star Wars Episode 4/5/6    Star Wars  158         True      1  
3495      Star Wars Episode 1    Star Wars  158         True      1

In [262]:
merged = df.merge(parent_theme, left_on ='parent_theme', right_on = 'name' )
merged.head(3)

set_num                        name_x  year  num_parts theme_name  \
0    00-1               Weetabix Castle  1970      471.0     Castle   
1    00-2  Weetabix Promotional House 1  1976        NaN   Building   
2    00-3  Weetabix Promotional House 2  1976        NaN   Building   

  parent_theme   id    name_y  is_licensed  
0     Legoland  411  Legoland        False  
1     Legoland  411  Legoland        False  
2     Legoland  411  Legoland        False

In [264]:
merged = merged.drop(columns = 'name_y', axis = 1)
merged.head(3)

set_num                        name_x  year  num_parts theme_name  \
0    00-1               Weetabix Castle  1970      471.0     Castle   
1    00-2  Weetabix Promotional House 1  1976        NaN   Building   
2    00-3  Weetabix Promotional House 2  1976        NaN   Building   

  parent_theme   id  is_licensed  
0     Legoland  411        False  
1     Legoland  411        False  
2     Legoland  411        False

# Task #1: What percentage of all licensed sets ever released were Star Wars Themed?

#### Licensed Star Wars sets / all licensed sets ever released
#### not all star wars theme name strings contain the exact string 'Star Wars'

In [266]:
licensed.shape[0]

1179

In [111]:
licensed = licensed.dropna(subset = ['set_num'])
licensed.shape[0]

1179

In [95]:
star_wars = licensed[licensed['parent_theme'] == 'Star Wars']

In [110]:
star_wars.shape[0]

609

In [109]:
licensed.shape[0]

1179

In [114]:
percent_of_starwars_sets = (star_wars.shape[0]/licensed.shape[0])*100
print(percent_of_starwars_sets)

51.653944020356235


# -- Answer: 51.65% --

# Task #2: In which year was Star Wars not the most popular licensed theme?

In [ ]:
licensed = merged[merged['is_licensed']== True]

In [165]:
licensed['count'] = 1
licensed.groupby(['year','parent_theme']).sum()

num_parts     id  is_licensed  count
year parent_theme                                                  
1999 Star Wars                    1384.0   2054           13     13
2000 Disney's Mickey Mouse         405.0   1940            5      5
     Star Wars                    2580.0   4108           26     26
2001 Harry Potter                 1284.0   2706           11     11
     Star Wars                    2949.0   2212           14     14
...                                  ...    ...          ...    ...
2017 Disney Princess               347.0   3474            6      6
     Minecraft                    5264.0   5193            9      9
     Pirates of the Caribbean     2286.0    263            1      1
     Star Wars                    7583.0   8690           55     55
     Super Heroes                13123.0  34704           72     72

[83 rows x 4 columns]

In [187]:
licensed

set_num                         name_x  year  num_parts  \
3493   10018-1                     Darth Maul  2001     1868.0   
3494   10019-1    Rebel Blockade Runner - UCS  2001        NaN   
3495   10026-1        Naboo Starfighter - UCS  2002        NaN   
3496   10030-1  Imperial Star Destroyer - UCS  2002     3115.0   
3497   10123-1                     Cloud City  2003      707.0   
...        ...                            ...   ...        ...   
11981   8410-1                      Swampfire  2010       22.0   
11982   8411-1                    ChromaStone  2010       21.0   
11983   8517-1                   Humungousaur  2010       14.0   
11984   8518-1                        Jet Ray  2010        NaN   
11985   8519-1                      Big Chill  2010       20.0   

                    theme_name parent_theme   id  is_licensed  count  
3493                 Star Wars    Star Wars  158         True      1  
3494   Star Wars Episode 4/5/6    Star Wars  158         True      1  
3495       Star Wars Episode 1    Star Wars  158         True      1  
3496   Star Wars Episode 4/5/6    Star Wars  158         True      1  
3497   Star Wars Episode 4/5/6    Star Wars  158         True      1  
...                        ...          ...  ...          ...    ...  
11981                   Ben 10       Ben 10  270         True      1  
11982                   Ben 10       Ben 10  270         True      1  
11983                   Ben 10       Ben 10  270         True      1  
11984                   Ben 10       Ben 10  270         True      1  
11985                   Ben 10       Ben 10  270         True      1  

[1179 rows x 9 columns]

In [201]:
licensed_summed = licensed.groupby(['year','parent_theme']).sum().reset_index()
licensed_summed.head(100)

maxdf = licensed_summed.sort_values('count', ascending = False).drop_duplicates('year')
maxdf.sort_values('year', ascending = False).drop(columns = ['num_parts','id','is_licensed','count'], axis=1)

year  parent_theme
82  2017  Super Heroes
76  2016     Star Wars
67  2015     Star Wars
59  2014     Star Wars
52  2013     Star Wars
47  2012     Star Wars
42  2011     Star Wars
36  2010     Star Wars
32  2009     Star Wars
28  2008     Star Wars
24  2007     Star Wars
20  2006     Star Wars
16  2005     Star Wars
12  2004     Star Wars
9   2003     Star Wars
6   2002     Star Wars
4   2001     Star Wars
2   2000     Star Wars
0   1999     Star Wars

# -- Answer: 2017 --

# Task 3: How many unique sets were released each year (1950-2017)?

In [246]:
m = merged.groupby(['year']).count()
m = m.drop(columns = ['set_num','name_x','num_parts','theme_name','parent_theme','id','is_licensed'])
pd.options.display.max_rows= 70
m

count
year       
1950      7
1953      4
1954     14
1955     28
1956     12
1957     21
1958     42
1959      4
1960      3
1961     17
1962     40
1963     18
1964     11
1965     10
1966     89
1967     21
1968     25
1969     69
1970     29
1971     45
1972     38
1973     68
1974     39
1975     31
1976     68
1977     92
1978     73
1979     82
1980     88
1981     79
1982     76
1983     57
1984     76
1985    139
1986    123
1987    209
1988     68
1989    114
1990     85
1991    106
1992    115
1993    111
1994    128
1995    128
1996    144
1997    194
1998    325
1999    300
2000    327
2001    342
2002    456
2003    418
2004    380
2005    332
2006    286
2007    321
2008    351
2009    403
2010    448
2011    504
2012    625
2013    608
2014    732
2015    680
2016    634
2017    474

# What is the average number of unique sets from each year (1950-2017)?

In [249]:
m.mean()

count    181.606061
dtype: float64